### PR TITLE
 FISH-766 Improper session synchronization of session map

### LINKED ISSUE #4280: Improper synchronization of session map
# Description #
----------
We are experiencing two kind of problems with SSO:

1. SSO expires prematurely (similar issue https://github.com/payara/Payara/issues/2881)
2. SSO never expires

I have done some debugging and found two possible minor issues with synchronization.

## Expected Outcome

No ConcurrentModificationExceptions during session expiration.

## Current Outcome
```
[2019-08-10T12:30:24.036+0200] [Payara 4.1] [WARNING] [AS-WEB-GLUE-00045] [javax.enterprise.web] [tid: _ThreadID=157 _ThreadName=SingleSignOnExpiration] [timeMillis: 1565433024036] [levelValue: 900] [[
  Caught exception during SingleSignOn expiration
java.util.ConcurrentModificationException
at java.util.HashMap$HashIterator.nextNode(HashMap.java:1445)
at java.util.HashMap$KeyIterator.next(HashMap.java:1469)
at com.sun.enterprise.security.web.GlassFishSingleSignOn.processExpires(GlassFishSingleSignOn.java:602)
at com.sun.enterprise.security.web.GlassFishSingleSignOn.run(GlassFishSingleSignOn.java:697)
at java.lang.Thread.run(Thread.java:748) 
```

## Steps to reproduce (Only for bug reports) 

Problem is reproducible with breakpoints at HASingleSignOn:170 and GlassfishSingleSignOn:697 with pausing only threads.

## Context (Optional)
HASingleSignOn calls cache.put without synchronized block on line 170. This may cause ConcurrentModificationException if some session expired and cleaner threed is actively looping over cache of sessions.

## Environment ##

- **Payara Version**: 4.1.1.171+

### PR DESCRIPTION
**Pull request by sgflt**
# Description
This is a bug fix of #4280 

# Important Info

### Blockers 

# Testing

### New tests

none
### Testing Performed

none
### Test suites executed

none

### Testing Environment

# Documentation

# Notes for Reviewers

Reviewer should check the synchronization was incorrect (asymetric) and now each write attempt is locked to prevent readers to receive ConcurrentModificationException.

### CODE REVIEW

**Comment by rdebusscher** [comment]
@sgflt  We are awaiting the signed CLA (https://github.com/payara/Payara/blob/master/PayaraCLA.pdf) otherwise we need to close this PR.

**Comment by smillidge** [comment]
We've got the CLA

**Comment by sgflt** [comment]
Just for minimal change. But why not. I can try to use ConcurrentHashMap.

**Comment by sgflt** [comment]
WIP: Replacing synchronized by ConcurrentHashMap is not so trivial. Some parts of code have different semantics that may break session management. Anyways I am keeping to dig deeper.

**Comment by sgflt** [comment]
What does `S1AS8 6155481 START` kind of comments mean?  Are they useful or just garbage?

**Comment by lprimak** [comment]
I think that's remanents of Sun / Oracle issue tracking before JIRA -type or source control integration

**Comment by lprimak** [comment]
> WIP: Replacing synchronized by ConcurrentHashMap is not so trivial. Some parts of code have different semantics that may break session management. Anyways I am keeping to dig deeper.

why not? I don't see any difference with that and synchronize, except synchronize has to be used in all parts of the code (even future parts) but not ConcurrentHashMap

File: appserver/web/web-glue/src/main/java/com/sun/enterprise/security/web/GlassFishSingleSignOn.java [hunk 1]
```diff
@@ -464,44 +461,38 @@ private void processExpires() {
...
-                while (it.hasNext()) {
-                    String key = it.next();
-                    SingleSignOnEntry sso = (SingleSignOnEntry) cache.get(key);
-                    if (sso.isEmpty() && sso.getLastAccessTime() < tooOld) {
-                        removals.add(key);
-                    }
+            this.cache.forEach((ssoId, sso) -> {
+                if (sso.isEmpty() && sso.getLastAccessTime() < tooOld) {
+                    removals.add(ssoId);
                 }
```
Comment (sgflt) [inline appserver/web/web-glue/src/main/java/com/sun/enterprise/security/web/GlassFishSingleSignOn.java:477]: There is a little change in behavior. 

Old implementation locked SSOs and created a list of expired SSOs. No one could create a new SSO.

New implementation uses what is available and does not block creation of a new SSO (which is not taken for expiration in gicen round). However probability of a new SSO being expired is so small so I think this change should be OK for a real world usage.

File: appserver/web/web-core/src/main/java/org/apache/catalina/session/StandardManager.java [hunk 2]
```diff
@@ -985,16 +982,14 @@ public void processExpires() {
 
         long timeNow = System.currentTimeMillis();
 
-        Session[] sessions = findSessions();
-        if (sessions != null) {
```
Comment (sgflt) [inline appserver/web/web-core/src/main/java/org/apache/catalina/session/StandardManager.java:989]: nullcheck for broken contract removed

File: appserver/web/web-core/src/main/java/org/apache/catalina/session/StandardManager.java [hunk 3]
```diff
@@ -882,21 +881,19 @@ public void stop(boolean isShutdown) throws LifecycleException {
         }
 
         // Expire all active sessions and notify their listeners
-        Session sessions[] = findSessions();
-        if (sessions != null) {
```
Comment (sgflt) [inline appserver/web/web-core/src/main/java/org/apache/catalina/session/StandardManager.java:886]: nullcheck for broken contract removed

File: appserver/web/web-core/src/main/java/org/apache/catalina/session/CookiePersistentManager.java [hunk 4]
```diff
@@ -111,8 +111,8 @@ public void clearSessions() {
     }
 
     @Override
-    public Session[] findSessions() {
-        return null;
```
Comment (sgflt) [inline appserver/web/web-core/src/main/java/org/apache/catalina/session/CookiePersistentManager.java:116]: broken contract

**Comment by dmatej** [comment]
> What does `S1AS8 6155481 START` kind of comments mean? Are they useful or just garbage?

S1AS8 means SunOne Application Server 8 which used same system as the Sun JDK, now dead.

**Comment by MeroRai** [comment]
jenkins test please

**Pull request merged**